scipy/_lib/tests/test__util.py

```diff
@@ -349,3 +352,46 @@ def mock_str():
     ]
 
     assert res == ref
+
+
+class TestLazywhere:
+    # Mixing floats and integers does not work in np.array_api, so don't try
+    n_arrays = strategies.integers(min_value=1, max_value=3)
+    rng_seed = strategies.integers(min_value=1000000000, max_value=9999999999)
+    dtype = strategies.sampled_from((np.float32, np.float64))
+    p = strategies.floats(min_value=0, max_value=1)
+    data = strategies.data()
+
+    @pytest.mark.filterwarnings('ignore::RuntimeWarning')  # overflows, etc.
+    @array_api_compatible
+    @given(n_arrays=n_arrays, rng_seed=rng_seed, dtype=dtype, p=p, data=data)
+    def test_basic(self, n_arrays, rng_seed, dtype, p, data, xp):
+        mbs = npst.mutually_broadcastable_shapes(num_shapes=n_arrays+1,
+                                                 min_side=0)
+        input_shapes, result_shape = data.draw(mbs)
+        cond_shape, *shapes = input_shapes
+        fillvalue = data.draw(npst.arrays(dtype=dtype, shape=tuple()))
+        arrays = [data.draw(npst.arrays(dtype=dtype, shape=shape))
+                  for shape in shapes]
+
+        def f(*args):
+            return sum(arg for arg in args)
+
+        def f2(*args):
+            return sum(arg for arg in args) / 2
+
+        rng = np.random.default_rng(rng_seed)
+        cond = rng.random(size=cond_shape) > p
+
+        xp_arrays = [xp.asarray(arr) for arr in arrays]
+        res1 = _lazywhere(xp.asarray(cond), xp_arrays, f, xp.asarray(fillvalue))
+        res2 = _lazywhere(xp.asarray(cond), xp_arrays, f, f2=f2)
+        type(res1) == type(res2) == type(xp.asarray([]))
+
+        # Ensure arrays are at least 1d to follow sane type promotion rules.
+        cond, fillvalue, *arrays = np.atleast_1d(cond, fillvalue, *arrays)
+        ref1 = np.where(cond, f(*arrays), fillvalue).reshape(result_shape)
+        ref2 = np.where(cond, f(*arrays), f2(*arrays)).reshape(result_shape)
+
+        assert_equal(np.asarray(res1), ref1)
+        assert_equal(np.asarray(res2), ref2)
```

Comment:
A `set_assert_equal` utility (like `set_assert_allclose` discussed in the `special` PR) could be useful here, to allow us to use the equivalent from the other libraries we want to test with.

Reply:
I meant to add a comment in the code about that to preempt this comment. But let's avoid scope creep in this PR. When `set_assert_equal` exists, we can use it, but I would prefer not to include it in this PR.

Reply:
That said, if you'd like to submit a PR for `set_assert_equal` (and perhaps `set_assert_allclose` separate from gh-19005) I would be willing to review it.

Reply:
Okay, I may be able to submit that PR in the next few days 👍 but please don't wait for me if this is ready to go in. In its current state, I think that you will need to add `@skip_if_array_api_gpu`, since `np.asarray` fails for CuPy arrays.

Reply:
`array_testing_namespace` seems like it might be a good idea. It may be worth it to look through the different submodules and see which assertions we would want to include. Or maybe we just include what we need for now and add to it progressively. Whereabouts in the file structure would this belong?

Reply:
Could [`assert_array_equal`](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_array_equal.html) be used here instead of `assert_equal`? It looks like CuPy has the former but not the latter.

Reply:
Sure.

Reply:
As `array_namespace` accepts arrays and returns a namespace with the appropriate Array API functions, `array_testing_namespace` would accept arrays and return a namespace with the appropriate testing functions. As `array_namespace` uses array_api_compat to give a consistent signature to existing functions, so `array_testing_namespace` would give a consistent signature to the various existing testing functions.

There are really not many array comparison testing functions to be handled. I really only need one to assess whether arrays are close (`assert_allclose`), equal (`assert_equal` or `assert_array_equal`, which don't seem different enough to exist separately aside from backward compatibility reasons), or ordered (`assert_array_less`).

So on second thought, rather than accepting arrays and returning a namespace, it might be better to just write three functions (e.g. `assert_equal`, `assert_close`, and `assert_less`) that accept two arrays, each of any type, and do the comparison as appropriate to the devices the  arrays live on, whatever that means. I assume that for two arrays on the GPU we would use CuPy's functions, two arrays accessible to the CPU would use NumPy or PyTorch test functions, and arrays split between the devices would move the data to one or the other before comparison.

Reply:
That sounds good to me - agreed that we only need those couple of functions.

Reply:
I think this suggestion is resolved, but I'll leave it open for others to refer to.